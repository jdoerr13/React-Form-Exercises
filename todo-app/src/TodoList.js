import React, { useState } from "react";
import { v4 as uuidv4 } from 'uuid';//provides a function to get a random id
import Todo from "./Todo"
import NewTodoForm from "./NewTodoForm";

//PARENT COMPONENT THAT WILL HOLD THE STATE: OR container component that manages the state of the shopping list and coordinates the addition of new items through the NewItemForm component.


const TodoList = () => {
  const [todos, setTodos] = useState([]);

  const addTodo = (task) => {
    const newTodo = { id: uuidv4(), task };// this is better with a string bc it is very unlikely to be an object
    setTodos(todos => [...todos, newTodo]);
  };
  

  const removeTodo = (id) => {
    setTodos(todos => todos.filter(todo => todo.id !== id));
  };

  const editTodo = (id, newTask) => {
    setTodos(todos => todos.map(todo =>
      todo.id === id ? { ...todo, task: newTask } : todo
    ));
  };

  return (
    <div>
      <h3>Task List</h3>
      <NewTodoForm addTodo={addTodo} />
      <div>
        {todos.map((todo) => 
        <Todo 
        key={todo.id}
        id={todo.id}
        task={todo.task}
        removeTodo={removeTodo}
        editTodo={editTodo}
        />)}
      </div>

    </div>
  )

}

export default TodoList;


