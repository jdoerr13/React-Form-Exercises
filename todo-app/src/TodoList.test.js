import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new task", function () {
  //select various pieces by queryByText & getByLabelText
  const { queryByText, getByPlaceholderText } = render(<TodoList />);
  const input = getByPlaceholderText("New task");
  const btn = queryByText("Add Todo");
  expect(queryByText('Chocolate Milk')).not.toBeInTheDocument();
  fireEvent.change(input, { target: { value: 'Chocolate Milk' } });//want input to change
  fireEvent.click(btn);//wee need this to submit the form
  expect(queryByText('Chocolate Milk')).toBeInTheDocument();
})


it("should remove a todo", function () {
  const { queryByText, getByPlaceholderText, getByText } = render(<TodoList />);
  
  // First, add a todo
  const input = getByPlaceholderText("New task");
  const addButton = queryByText("Add Todo");
  fireEvent.change(input, { target: { value: 'Chocolate Milk' } });
  fireEvent.click(addButton);

   // Verify todo is added
   expect(queryByText('Chocolate Milk')).toBeInTheDocument();

   // Step 2: Remove the todo
   const deleteButton = getByText('X'); // Assuming 'X' is the text on your delete button
   fireEvent.click(deleteButton);
 
   // Step 3: Verify todo is removed
   expect(queryByText('Chocolate Milk')).not.toBeInTheDocument();
 });