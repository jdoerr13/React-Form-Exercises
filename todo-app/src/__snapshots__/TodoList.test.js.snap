// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Task List
    </h3>
    <form>
      <input
        placeholder="New task"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
